### DISCUSSION TITLE
`argon2*` functions don't accept version. Instead it is hardcoded as `ARGON2_VERSION_NUMBER`. Should this lib support multiple versions?

### DISCUSSION THREAD

**Discussion by KOLANICH** [jedisct1/libsodium#1284] [Ideas]
It may cause issues if/when this lib has to have the hash function version upgraded. If a certain version of Argon2 was used to hash secrets, the same version would have to be used when the same result is expected. This way, upgrading Argon2 version in this lib would cause breakage of the software depending on it. The proposed way to deal with it is

1. add `version` field to `argon2_context`, and initialize it with `ARGON2_VERSION_NUMBER` by default.
2. deprecate the current set of functions, leave `argon2_ctx` or maybe a new wrapper around it as a recommended method of doing hashing.
3. state, that the current set of functions use the current value of `ARGON2_VERSION_NUMBER`. Put it into `ARGON2_LEGACY_API_VERSION_NUMBER`, while `ARGON2_VERSION_NUMBER` is free to be incremented.
4. implement all the versions of the functions and dispatch based on `version` field.

Users using the legacy API are free to do it. Users using the new API are expected to set `version` field of the context object to the version they expect/store it somewhere for future use.

**Comment by jedisct1** [comment]
Use the documented, high-level functions: `crypto_pwhash_*()`.

**Reply by KOLANICH** [reply]
Thanks for the info.
